### PR TITLE
AUTO-14 OS-1882 - exclude cinder volume from backup 

### PR DESCRIPTION
**Pull request by SuhasTA4**
OS-1882 - New test case to exclude cinder volume from taking backup.

### CODE REVIEW

**Comment by SuhasTA4** [comment]
[[Report]]
image

**Comment by SuhasTA4** [comment]
[[bootdisk_exclusion Report]]
image

File: tempest/api/workloadmgr/restore/test_exclude_cinder_volume.py [hunk 1]
```diff
@@ -0,0 +1,266 @@
...
+        fip = self.get_floating_ips()
+        LOG.debug("\nAvailable floating ips are {}: \n".format(fip))
+        self.set_floating_ip(str(fip[0]), vm_id, floatingip_cleanup=cleanup)
+        return(fip[0])
+
+
+    def data_ops_on_mount_point(self, flo_ip, mount_point, file_count):
+        global volumes
+        ssh = self.SshRemoteMachineConnectionWithRSAKey(str(flo_ip))
+        stdin, stdout, stderr = ssh.exec_command(
```
Comment (savitha-p) [inline tempest/api/workloadmgr/restore/test_exclude_cinder_volume.py]: Use already existing method install_qemu() from base.py instead of repeating it here.

File: tempest/api/workloadmgr/restore/test_exclude_cinder_volume.py [hunk 2]
```diff
@@ -0,0 +1,266 @@
...
+            LOG.debug("Assigned floating IP : " + str(floating_ip_1))
+            time.sleep(20)
+
+            #add some data to volume attached
+            mount_point = ["mount_data_AFolder"]
+            self.data_ops_on_mount_point(floating_ip_1, mount_point[0], 3)
+            LOG.debug("Created disk and mounted the attached volume")
+
+            #add metadata to the volume...
+            custom_metadata = {"exclude_from_backup": "True"
```
Comment (savitha-p) [inline tempest/api/workloadmgr/restore/test_exclude_cinder_volume.py]: Move this to tvaultconf.py

File: tempest/api/workloadmgr/restore/test_exclude_cinder_volume.py [hunk 3]
```diff
@@ -0,0 +1,266 @@
...
+                        }
+
+            body = self.volumes_client.update_volume_metadata(
```
Comment (savitha-p) [inline tempest/api/workloadmgr/restore/test_exclude_cinder_volume.py]: Create a reusable method in base.py for this.

File: tempest/api/workloadmgr/restore/test_exclude_cinder_volume.py [hunk 4]
```diff
@@ -0,0 +1,266 @@
...
+
+            if (snapshot_status == "available"):
+                LOG.debug("Full snapshot created.")
+                reporting.add_test_step("Create full snapshot", tvaultconf.PASS)
+                tests[0][1] = 1
+            else:
+                reporting.add_test_step("Create full snapshot", tvaultconf.FAIL)
+                raise Exception("Full snapshot creation failed.")
+
+            reporting.test_case_to_write()
```
Comment (savitha-p) [inline tempest/api/workloadmgr/restore/test_exclude_cinder_volume.py]: Why is snapshot creation considered a separate test case ? Restore is the test case here I suppose.

File: tempest/api/workloadmgr/restore/test_exclude_cinder_volume.py [hunk 5]
```diff
@@ -0,0 +1,266 @@
...
+            cmd = "sudo mount "+volumes[0]+"1 "+mount_point[0]+" "
+            LOG.debug("Command to mount volume" + str(cmd))
+            stdin, stdout, stderr = ssh.exec_command(
+            cmd)
+            time.sleep(5)
+            while not stdout.channel.exit_status_ready():
+                time.sleep(3)
+                max_retries += 1
+                if max_retries == 10:
+                    reporting.add_test_step("Max retries exceeded to get command output.", tvaultconf.FAIL)
```
Comment (savitha-p) [inline tempest/api/workloadmgr/restore/test_exclude_cinder_volume.py]: Add more details as to what command.

File: tempest/api/workloadmgr/restore/test_exclude_cinder_volume.py [hunk 6]
```diff
@@ -0,0 +1,266 @@
...
+                    tests[1][1] = 1
+                else:
+                    reporting.add_test_step("Different error occurred than expected.", tvaultconf.FAIL)
+                    raise Exception("Different error occurred while performing mount operation.")
+            else:
+                reporting.add_test_step("Different error occurred than expected.", tvaultconf.FAIL)
+                raise Exception("Mount command successful, test case failed.")
+
+            #close ssh connection
+            ssh.close()
```
Comment (savitha-p) [inline tempest/api/workloadmgr/restore/test_exclude_cinder_volume.py]: In case of any exception, ssh connection does not get closed. Need to handle it.

File: tempest/api/workloadmgr/restore/test_exclude_cinder_volume.py [hunk 7]
```diff
@@ -0,0 +1,266 @@
...
+
+
+        except Exception as e:
+            for test in tests:
```
Comment (savitha-p) [inline tempest/api/workloadmgr/restore/test_exclude_cinder_volume.py:423]: This needs to be moved to finally block I suppose.

File: tempest/api/workloadmgr/restore/test_exclude_bootdisk_backup.py [hunk 8]
```diff
@@ -0,0 +1,367 @@
...
+
+    def data_ops_for_bootdisk(self, flo_ip, data_dir_path, file_count):
+        ssh = self.SshRemoteMachineConnectionWithRSAKey(str(flo_ip))
+        self.install_qemu(ssh)
+        self.addCustomfilesOnLinuxVM(ssh, data_dir_path, file_count)
+        ssh.close()
+
+
+    @decorators.attr(type='workloadmgr_cli')
+    def test_exclude_bootdisk_backup(self):
```
Comment (savitha-p) [inline tempest/api/workloadmgr/restore/test_exclude_bootdisk_backup.py]: Rename the function name to exclude_bootdisk_backup() so that it does not get called during test suite execution.

File: tempest/api/workloadmgr/restore/test_exclude_bootdisk_backup.py [hunk 9]
```diff
@@ -0,0 +1,367 @@
...
+            if error:
+                err_msg = ["No such file or directory"]
+                result = all(x in error for x in err_msg)
+                if (result):
+                    reporting.add_test_step("Search files on restored vm instance",
+                                            tvaultconf.PASS)
+                    reporting.set_test_script_status(tvaultconf.PASS)
+                    tests[1][1] = 1
+                else:
+                    reporting.add_test_step("Different error occurred.",
```
Comment (nasharatm) [inline tempest/api/workloadmgr/restore/test_exclude_bootdisk_backup.py]: Please reword this to reflect correct step failure. You can use same text from line #222.

File: tempest/api/workloadmgr/restore/test_exclude_bootdisk_backup.py [hunk 10]
```diff
@@ -0,0 +1,367 @@
...
+            output = stdout.read().decode('utf-8').strip()
+            error = stderr.read().decode('utf-8').strip()
+
+            LOG.debug("Output: " + str(output))
+            LOG.debug("Error: " + str(error))
+
+            if error:
+                err_msg = ["No such file or directory"]
+                result = all(x in error for x in err_msg)
+                if (result):
```
Comment (nasharatm) [inline tempest/api/workloadmgr/restore/test_exclude_bootdisk_backup.py:222]: Is error expected here? and what would be captured in result?

File: tempest/api/workloadmgr/restore/test_exclude_bootdisk_backup.py [hunk 11]
```diff
@@ -0,0 +1,367 @@
...
+
+            #write the test step into test case - filesearch. 
+            #Remove metadata...
+            reporting.test_case_to_write()
+
+            reporting.add_test_script(tests[2][0])
+
+            
+            #add some more data and remove metadata for verification..
+            ssh = self.SshRemoteMachineConnectionWithRSAKey(str(floating_ip_1))
```
Comment (nasharatm) [inline tempest/api/workloadmgr/restore/test_exclude_bootdisk_backup.py]: This block is repetitive. Can you create single function for the same?

File: tempest/api/workloadmgr/restore/test_exclude_cinder_volume.py [hunk 12]
```diff
@@ -0,0 +1,266 @@
...
+
+            reporting.add_test_script(tests[0][0])
+
+            #create key_pair value 
+            kp = self.create_key_pair(
+                tvaultconf.key_pair_name, keypair_cleanup=True)
+            LOG.debug("Key_pair : " + str(kp))
+
+            #create vm
+            vm_id = self.create_vm(key_pair=kp, vm_cleanup=False)
```
Comment (nasharatm) [inline tempest/api/workloadmgr/restore/test_exclude_cinder_volume.py]: we should set vm_cleanup = true so that it can be cleaned up after test execution. That way we can remove specific code to handle vm deletion in finally.

**Review by nasharatm — CHANGES_REQUESTED** [review]

File: tempest/api/workloadmgr/restore/test_exclude_cinder_volume.py [hunk 13]
```diff
@@ -0,0 +1,266 @@
...
+
+    @decorators.attr(type='workloadmgr_cli')
+    def test_exclude_cinder_volume_backup(self):
+        try:
+            deleted = 0
+            global volumes
+            ## VM and Workload ###
+            tests = [['tempest.api.workloadmgr.restore.add_metadata_to_exclude_cinder_volume_from_backup',
+                      0],
+                     ['tempest.api.workloadmgr.restore.check_volume_status_after_restore',
```
Comment (savitha-p) [inline tempest/api/workloadmgr/restore/test_exclude_cinder_volume.py]: Rename the test case to "selective restore of full snapshot with exclude cinder volume metadata"

File: tempest/api/workloadmgr/restore/test_exclude_cinder_volume.py [hunk 14]
```diff
@@ -0,0 +1,266 @@
...
+        ssh.close()
+
+
+    @decorators.attr(type='workloadmgr_cli')
+    def test_exclude_cinder_volume_backup(self):
+        try:
+            deleted = 0
+            global volumes
+            ## VM and Workload ###
+            tests = [['tempest.api.workloadmgr.restore.add_metadata_to_exclude_cinder_volume_from_backup',
```
Comment (savitha-p) [inline tempest/api/workloadmgr/restore/test_exclude_cinder_volume.py]: Metadata addition is an openstack feature, not required to be added as test case.

File: tempest/api/workloadmgr/restore/test_exclude_bootdisk_backup.py (hunk 10 above)
Comment (SuhasTA4) [reply tempest/api/workloadmgr/restore/test_exclude_bootdisk_backup.py:222]: Error is expected and error coming as "No such file or directory" because we are checking file status on restored instance. And it is expected that file should not be present.

File: tempest/api/workloadmgr/base.py [hunk 15]
```diff
@@ -553,6 +553,21 @@ def detach_volumes(self, volumes):
...
+        body = ""
+        try:
+            body = self.volumes_client.update_volume_metadata(
+                volume_id, metadata_tag_name)['metadata']
+
+        except Exception as e:
+            LOG.error("Exception" + str(e))
+            pass
+
+        return body
```
Comment (savitha-p) [inline tempest/api/workloadmgr/base.py:569]: in case of exception, this return statement will break

**Comment by SuhasTA4** [comment]
test_exclude_cinder_volume result:

image

File: tempest/api/workloadmgr/restore/test_exclude_cinder_volume.py [hunk 16]
```diff
@@ -0,0 +1,444 @@
...
+                wid,
+                snapshot_id,
+                restore_name=tvaultconf.restore_name,
+                restore_cleanup=True,
+                instance_details=payload['instance_details'],
+                network_details=payload['network_details'])
+
+            self.wait_for_snapshot_tobe_available(wid, snapshot_id)
+            if (self.getRestoreStatus(wid, snapshot_id, restore_id_1) == "available"):
+                reporting.add_test_step("Selective restore", tvaultconf.PASS)
```
Comment (savitha-p) [inline tempest/api/workloadmgr/restore/test_exclude_cinder_volume.py]: Include snapshot used for restore in the report for each restore

File: tempest/api/workloadmgr/restore/test_exclude_cinder_volume.py [hunk 17]
```diff
@@ -0,0 +1,444 @@
...
+
+            if (snapshot_status == "available"):
+                LOG.debug("Full snapshot created.")
+                reporting.add_test_step("Create full snapshot", tvaultconf.PASS)
+            else:
+                reporting.add_test_step("Create full snapshot", tvaultconf.FAIL)
+                raise Exception("Full snapshot creation failed.")
+
+            # take incremental snapshot
+            incr_snapshot_id = self.create_snapshot(wid, is_full=False)
```
Comment (savitha-p) [inline tempest/api/workloadmgr/restore/test_exclude_cinder_volume.py]: What is the difference between full and incremental snapshot here? There is no data update done on the volume side.

Comment (SuhasTA4) [reply tempest/api/workloadmgr/restore/test_exclude_cinder_volume.py]: It was there earlier but then I removed it because it was not serving any purpose as it is (incr_snapshot_id) only used in One-Click restore and there also we are not calculating any checksum to validate. 
We are just checking mount command status and for that purpose we don't need any additional data on the volume. 

Please correct me if I am wrong.  If you think we should still add more data before incremental snapshot, I will add those code block again.

File: tempest/api/workloadmgr/restore/test_exclude_cinder_volume.py [hunk 18]
```diff
@@ -0,0 +1,449 @@
...
+                wid,
+                snapshot_id,
+                restore_name=tvaultconf.restore_name,
+                restore_cleanup=True,
+                instance_details=payload['instance_details'],
+                network_details=payload['network_details'])
+
+            self.wait_for_snapshot_tobe_available(wid, snapshot_id)
+            if (self.getRestoreStatus(wid, snapshot_id, restore_id_1) == "available"):
+                reporting.add_test_step("Selective restore of snapshot id = {} ".format(
```
Comment (savitha-p) [inline tempest/api/workloadmgr/restore/test_exclude_cinder_volume.py]: Including snapshot id in report will not help , instead u need to include type of snapshot. Like selective restore of full snapshot. That will help team identify which snapshot is restored.

**Review by savitha-p — APPROVED** [review]

**Review by nasharatm — APPROVED** [review]
